README.md
[![License](https://img.shields.io/npm/l/feathers-redis-cache.svg)](https://www.npmjs.com/package/feathers-redis-cache)
[![NPM](https://img.shields.io/npm/v/feathers-redis-cache.svg)](https://www.npmjs.com/package/feathers-redis-cache)

##### This repository is a fork of [feathers-hooks-rediscache](https://github.com/idealley/feathers-hooks-rediscache), with the following changes:
- refactor and simplified API and source code (details bellow).
- support of feathers pagination hook, when it's enabled or disabled per the same endpoint.
- key always generated in before hook.
- option to pass custom logger function to log errors.
- option to pass custom `/cache` prefix for redis routs.
- option to pass custom `cacheKey(hook)` function in before hook, to customize key generation.
- implementing `.purge()` hook, so you can place in after create/patch/update/remove hooks to purge entire group and avoid having keys with corrupted values in the cache.
- remove `hookCache` now to set custom expiration date you need to pass it as an option in after hook.
- remove `redisCache` from config.
- remove `parseNestedRoutes`, `removePathFromCacheKey`, `defaultDuration` (by default it will expire in 24hr), `immediateCacheKey` options.

### Installation

```
  yarn add feathers-redis-cache
```    
```
  npm install feathers-redis-cache
```    

## Purpose
The purpose of these hooks is to provide redis caching for APIs endpoints. Using redis is a very good option for clustering your API. As soon as a request is cached it is available to all the other nodes in the cluster, which is not true for usual in memory cache as each node has its own memory allocated. This means that each node has to cache all requests individually.

Each request to an endpoint can be cached. Route variables and params are cached on a per request base. If a param to call is set to true and then to false two responses will be cached.

The cache can be purged for an individual route, but also for a group of routes. This is very useful if you have an API endpoint that creates a list of articles, and an endpoint that returns an individual article. If the article is modified, the list of articles should, most likely, be purged as well. This can be done by calling one endpoint.

### Routes exemples
In the same fashion if you have many variants of the same endpoint that return similar content based on parameters you can bust the whole group as well:

```js
'/articles' // list
'/articles/article' //individual item
'/articles/article?markdown=true' // variant
```
#### Clearing cache
These are all listed in a redis list under `group-articles` and can be busted by calling `/cache/clear/group/articles`. All urls keys will be purged.

You can also purge single cached paths as by doing GET requests on 
```js
'/cache/clear/single/articles'
'/cache/clear/single/articles/article'
'/cache/clear/single/articles/article?markdown=true' // works with query strings too
```

or purge all by calling `/cache/clear/all`

It was meant to be used over **_HTTP_**, not yet tested with sockets.

### Configuration
#### Redis
To configure the redis connection the feathers configuration system can be used.
```js
//config/default.json
{
  "host": "localhost",
  "port": 3030,
  "redis": {
    "host": "my-redis-service.example.com",
    "port": 1234
  }
}
```
* if no config is provided, default config from the [redis module](https://github.com/NodeRedis/node_redis) is used

## Available hooks
More details and example use bellow

* `hooks.before(options)` - retrieves the data from redis
* `hooks.after(options)` - cache the data to redis
* `hooks.purge()` - purge cache from redis

#### options properties (all props are optional)

##### cacheKey(context: `feathers-context`): `string`
In case if you want to use custom function to modify key name, you need to pass the same function in before and after hooks.

##### expiration: `number`
Time in seconds when to expire the key, this option need to be passed in after hook, if you won't pass it, default value of 1 day will be used.

##### env: `string`
The default environment is production, but it is annoying when running test as the hooks output information to the console. Therefore if you use this option, you can set `test` as an environment and the hooks will not output anything to the console. This is useful for CI or CLI.

Available routes:
```js
'/cache/clear/all' // clears the whole cache
'/cache/clear/single/:target' // clears a single route if you want to purge a route with params just adds them target?param=1
'/cache/clear/group/:target' // clears a group
```

## Complete Example

Here's an example of a Feathers server that uses `feathers-redis-cache`.

```js
const feathers = require('feathers');
const rest = require('feathers-rest');
const hooks = require('feathers-hooks');
const bodyParser = require('body-parser');
const errorHandler = require('feathers-errors/handler');
const redisCache = require('feathers-redis-cache');

// Initialize the application
const app = feathers()
  .use(bodyParser.json())
  .use(bodyParser.urlencoded({ extended: true }))
  .configure(rest())
  .configure(hooks())
  // errorLogger is function for logging errors
  // if not passed console.error will bbe used
  .configure(redisCache.client({ errorLogger: logger.error }))
  // you can change cache path prefix by passing `pathPrefix` option
  // if not passed default prefix '/cache' will be used
  .configure(redisCache.services({ pathPrefix: '/cache' }))
  .use(errorHandler());

app.listen(3030);

console.log('Feathers app started on 127.0.0.1:3030');
```

Add hooks on the routes that need caching
```js
//services/<service>.hooks.js

const redisCache = require('feathers-redis-cache');


module.exports = {
  before: {
    all: [],
    find: [redisCache.before()],
    get: [redisCache.before()],
    create: [],
    update: [],
    patch: [],
    remove: []
  },

  after: {
    all: [],
    find: [redisCache.after({ expiration: 3600 * 24 * 7 })],
    get: [redisCache.after({ expiration: 3600 * 24 * 7 })],
    create: [redisCache.purge()],
    update: [redisCache.purge()],
    patch: [redisCache.purge()],
    remove: [redisCache.purge()]
  },

  error: {
    all: [],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  }
};
```

You can also skip cache hook by passing `hook.params.$skipCacheHook = true`
You can also disable redis-cache hooks and service by passing env. variable `DISABLE_REDIS_CACHE=true`

## TODO:
- TS definitions
- test cases
- option in after hook to set limit of keys per group
## License

Copyright (c) 2018

Licensed under the [MIT license](LICENSE).


CHANGELOG.md
## [1.0.12](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.11...v1.0.12) (2019-05-03)


### Bug Fixes

* **readme:** minor updates ([87ff469](https://github.com/sarkistlt/feathers-redis-cache/commit/87ff469))

## [1.0.11](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.10...v1.0.11) (2019-05-03)


### Bug Fixes

* **readme:** typo fix ([99ad0f8](https://github.com/sarkistlt/feathers-redis-cache/commit/99ad0f8))

## [1.0.10](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.9...v1.0.10) (2019-05-03)


### Bug Fixes

* **readme:** typo fix ([6203bf2](https://github.com/sarkistlt/feathers-redis-cache/commit/6203bf2))

## [1.0.9](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.8...v1.0.9) (2019-05-03)


### Bug Fixes

* **readme:** add labels ([ad93576](https://github.com/sarkistlt/feathers-redis-cache/commit/ad93576))

## [1.0.8](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.7...v1.0.8) (2019-05-03)


### Bug Fixes

* **gitignore:** typo fix ([6018510](https://github.com/sarkistlt/feathers-redis-cache/commit/6018510))

## [1.0.7](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.6...v1.0.7) (2019-05-03)


### Bug Fixes

* **force update:** force patch release to update readme on npm ([9e19255](https://github.com/sarkistlt/feathers-redis-cache/commit/9e19255))

## [1.0.6](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.5...v1.0.6) (2019-04-29)


### Bug Fixes

* **ObjectID:** parse objectID if it was passed as an query ([921b188](https://github.com/sarkistlt/feathers-redis-cache/commit/921b188))

## [1.0.5](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.4...v1.0.5) (2019-04-29)


### Bug Fixes

* **cacheKey:** add qs option arrayFormat: 'comma' ([5969d21](https://github.com/sarkistlt/feathers-redis-cache/commit/5969d21))

## [1.0.4](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.3...v1.0.4) (2019-04-29)


### Bug Fixes

* **cacheKey:** change format of parsed url ([3fb779a](https://github.com/sarkistlt/feathers-redis-cache/commit/3fb779a))

## [1.0.3](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.2...v1.0.3) (2019-04-26)


### Bug Fixes

* **implement disable mech.:** disable hooks and services by passing DISABLE_REDIS_CACHE ([9830514](https://github.com/sarkistlt/feathers-redis-cache/commit/9830514))

## [1.0.2](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.1...v1.0.2) (2019-04-25)


### Bug Fixes

* **modules:** import error, resolved ([bc5ff7a](https://github.com/sarkistlt/feathers-redis-cache/commit/bc5ff7a))

## [1.0.1](https://github.com/sarkistlt/feathers-redis-cache/compare/v1.0.0...v1.0.1) (2019-04-25)


### Bug Fixes

* **patch:** force new patch version ([f570f8d](https://github.com/sarkistlt/feathers-redis-cache/commit/f570f8d))

# 1.0.0 (2019-04-25)


### Bug Fixes

* **build error:** fix error with getting configs ([b71a16d](https://github.com/sarkistlt/feathers-redis-cache/commit/b71a16d))

# 1.0.0 (2019-04-25)


### Bug Fixes

* **build error:** fix error with getting configs ([b71a16d](https://github.com/sarkistlt/feathers-redis-cache/commit/b71a16d))


src/client.ts
/* eslint-disable no-console */
import redis from 'redis';
import chalk from 'chalk';

const { DISABLE_REDIS_CACHE } = process.env;

export default (options: any = {}) => {
  const errorLogger = options.errorLogger || console.error;
  const retryInterval = options.retryInterval || 5000;

  if (DISABLE_REDIS_CACHE) {
    return () => {};
  }

  return function client() {
    const app = this;
    const config = app.get('redis') || {};

    try {
      const redisOptions = {
        ...config,
        retry_strategy: () => {
          app.set('redisClient', undefined);

          console.log(`${chalk.yellow('[redis]')} not connected`);

          return retryInterval;
        }
      };
      const client = redis.createClient(redisOptions);

      app.set('redisClient', client);

      client.on('ready', () => {
        app.set('redisClient', client);

        console.log(`${chalk.green('[redis]')} connected`);
      });
    } catch (err) {
      errorLogger(err);
      app.set('redisClient', undefined);
    }

    return this;
  };
}


src/services.ts
import async from 'async';

const { DISABLE_REDIS_CACHE } = process.env;
const HTTP_OK = 200;
const HTTP_NO_CONTENT = 204;
const HTTP_SERVER_ERROR = 500;
const HTTP_NOT_FOUND = 404;

const serviceClearSingle = {
  setup(app, path) {
    this.app = app;
    this.path = path;
  },
  async find(params) {
    const client = this.app.get('redisClient');
    const { target } = params.query;

    if (!target) {
      return {
        message: 'You must provide key',
        status: HTTP_NOT_FOUND
      };
    }

    if (!client) {
      return {
        message: 'Redis unavailable',
        status: HTTP_SERVER_ERROR
      };
    }

    return new Promise((resolve) => {
      client.get(target, (err, reply) => {
        if (err) {
          return resolve({ message: 'something went wrong' + err.message });
        }

        if (!reply) {
          return resolve({
            message: `cache already cleared for key ${target}`,
            status: HTTP_NO_CONTENT
          });
        }

        client.del(target, (err, reply) => {
          if (err) {
            return resolve({ message: 'something went wrong' + err.message });
          }

          if (!reply) {
            return resolve({
              message: `cache already cleared for key ${target}`,
              status: HTTP_NO_CONTENT
            });
          }

          resolve({
            message: `cache cleared for key ${target}`,
            status: HTTP_OK
          });
        });
      });
    });
  },
};
const serviceClearGroup = {
  setup(app, path) {
    this.app = app;
    this.path = path;
  },
  async find(params) {
    const client = this.app.get('redisClient');
    const { target } = params.query;

    if (!client) {
      return {
        message: 'Redis unavailable',
        status: HTTP_SERVER_ERROR
      };
    }

    return new Promise((resolve) => {
      client.lrange(`group-${target}`, 0, -1, (err, reply) => {
        if (err) {
          return resolve({ message: 'something went wrong' + err.message });
        }

        // If the list/group existed and contains something
        if (!reply || !Array.isArray(reply) || reply.length <= 0) {
          return resolve({
            message: `cache already cleared for the group key: ${target}`,
            status: HTTP_NO_CONTENT
          });
        }

        async.eachOfLimit(reply, 10, async.asyncify(async (key) => {
          return new Promise((res) => {
            client.del(key, (err, reply) => {
              if (err) {
                return res({ message: 'something went wrong' + err.message });
              }

              if (!reply) {
                return res({
                  message: `cache already cleared for key ${target}`,
                  status: HTTP_NO_CONTENT
                });
              }

              res({
                message: `cache cleared for key ${target}`,
                status: HTTP_OK
              });
            });
          });
        }), (err) => {
          if (err) {
            return resolve({ message: 'something went wrong' + err.message });
          }
          resolve({
            message: `cache cleared for the group key: ${target}`,
            status: HTTP_OK
          });
        });
      });
    });
  },
};
const serviceClearAll = {
  setup(app, path) {
    this.app = app;
    this.path = path;
  },
  async find() {
    const client = this.app.get('redisClient');

    if (!client) {
      return {
        message: 'Redis unavailable',
        status: HTTP_SERVER_ERROR
      };
    }

    return new Promise((resolve) => {
      client.flushdb(() => {
        resolve({
          message: 'Cache cleared',
          status: HTTP_OK
        });
      });
    });
  },
};

export default (options: any = {}) => {
  const pathPrefix = options.pathPrefix || '/cache';

  return function () {
    const app = this;

    if (!DISABLE_REDIS_CACHE) {
      app.use(`${pathPrefix}/clear/single`, serviceClearSingle);
      app.use(`${pathPrefix}/clear/group`, serviceClearGroup);
      app.use(`${pathPrefix}/clear/all`, serviceClearAll);
    }
  };
}



src/hooks.ts
/* eslint-disable no-console */
import moment from 'moment/moment';
import chalk from 'chalk';
import qs from 'qs';
import async from 'async';

const { DISABLE_REDIS_CACHE } = process.env;
const HTTP_OK = 200;
const HTTP_NO_CONTENT = 204;
const HTTP_SERVER_ERROR = 500;
const defaults = {
  defaultExpiration: 3600 * 24
};

function cacheKey(hook) {
  const q = hook.params.query || {};
  const p = hook.params.paginate === false ? 'disabled' : 'enabled';
  let path = `pagination-hook:${p}::${hook.path}`;

  if (hook.id) {
    path += `/${hook.id}`;
  }

  if (Object.keys(q).length > 0) {
    path += `?${qs.stringify(JSON.parse(JSON.stringify(q)), { encode: false, arrayFormat: 'comma' })}`;
  }

  return path;
}

export default {
  before(passedOptions) {
    if (DISABLE_REDIS_CACHE) {
      return hook => hook;
    }

    return function (hook) {
      try {
        if (hook && hook.params && hook.params.$skipCacheHook) {
          return Promise.resolve(hook);
        }

        return new Promise(resolve => {
          const client = hook.app.get('redisClient');
          const options = { ...defaults, ...passedOptions };

          if (!client) {
            return resolve(hook);
          }

          const path = typeof options.cacheKey === 'function' ?
            options.cacheKey(hook) :
            cacheKey(hook);

          hook.params.cacheKey = path;

          client.get(path, (err, reply) => {
            if (err) {
              return resolve(hook);
            }

            if (reply) {
              const data = JSON.parse(reply);

              if (!data || !data.expiresOn || !data.cache) {
                return resolve(hook);
              }

              const duration = moment(data.expiresOn).format('DD MMMM YYYY - HH:mm:ss');

              hook.result = data.cache;
              hook.params.$skipCacheHook = true;

              if (options.env !== 'test') {
                console.log(`${chalk.cyan('[redis]')} returning cached value for ${chalk.green(path)}.`);
                console.log(`> Expires on ${duration}.`);
              }

              return resolve(hook);
            }

            return resolve(hook);
          });
        });
      } catch (err) {
        console.error(err);
        return Promise.resolve(hook);
      }
    };
  },
  after(passedOptions) {
    if (DISABLE_REDIS_CACHE) {
      return hook => hook;
    }

    return function (hook) {
      try {
        if (
          hook
          && hook.params
          && hook.params.$skipCacheHook
        ) {
          return Promise.resolve(hook);
        }

        if (!hook.result) {
          return Promise.resolve(hook);
        }

        return new Promise((resolve) => {
          const client = hook.app.get('redisClient');
          const options = { ...defaults, ...passedOptions };
          const duration = options.expiration || options.defaultExpiration;
          const { cacheKey: path } = hook.params;
          const group = hook.path ? `group-${hook.path}` : '';

          if (!client) {
            return resolve(hook);
          }

          client.set(path, JSON.stringify({
            cache: hook.result,
            expiresOn: moment().add(moment.duration(duration, 'seconds')),
            group,
          }));
          client.expire(path, duration);
          client.rpush(group, path);

          if (options.env !== 'test') {
            console.log(`${chalk.cyan('[redis]')} added ${chalk.green(path)} to the cache.`);
            console.log(`> Expires in ${moment.duration(duration, 'seconds').humanize()}.`);
          }

          resolve(hook);
        });
      } catch (err) {
        console.error(err);
        return Promise.resolve(hook);
      }
    };
  },
  purge() {
    if (DISABLE_REDIS_CACHE) {
      return hook => hook;
    }

    return function (hook) {
      try {
        if (
          hook
          && hook.params
          && hook.params.$skipCacheHook
        ) {
          return Promise.resolve(hook);
        }

        return new Promise((resolve) => {
          const client = hook.app.get('redisClient');
          const target = hook.path;

          if (!client) {
            return {
              message: 'Redis unavailable',
              status: HTTP_SERVER_ERROR
            };
          }

          client.lrange(`group-${target}`, 0, -1, (err, reply) => {
            if (err) {
              return resolve(hook);
            }

            if (!reply || !Array.isArray(reply) || reply.length <= 0) {
              return resolve(hook);
            }

            async.eachOfLimit(reply, 10, async.asyncify(async (key) => {
              return new Promise((res) => {
                client.del(key, (err, reply) => {
                  if (err) {
                    return res({ message: 'something went wrong' + err.message });
                  }

                  if (!reply) {
                    return res({
                      message: `cache already cleared for key ${target}`,
                      status: HTTP_NO_CONTENT
                    });
                  }

                  res({
                    message: `cache cleared for key ${target}`,
                    status: HTTP_OK
                  });
                });
              });
            }), () => resolve(hook));
          });
        });
      } catch (err) {
        console.error(err);
        return Promise.resolve(hook);
      }
    };
  },
};
